beacon-chain/types/active_state.go

```diff
@@ -0,0 +1,212 @@
+func (a *ActiveState) Proto() *pb.ActiveState {
+	return a.data
+}
+
+// Marshal encodes active state object into the wire format.
+func (a *ActiveState) Marshal() ([]byte, error) {
+	return proto.Marshal(a.data)
+}
+
+// Hash serializes the active state object then uses
+// blake2b to hash the serialized object.
+func (a *ActiveState) Hash() ([32]byte, error) {
+	data, err := proto.Marshal(a.data)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	var hash [32]byte
+	h := blake2b.Sum512(data)
+	copy(hash[:], h[:32])
+	return hash, nil
+}
+
+// PendingAttestations returns attestations that have not yet been processed.
+func (a *ActiveState) PendingAttestations() []*pb.AttestationRecord {
+	return a.data.PendingAttestations
+}
+
+// RecentBlockHashes returns the most recent 2*EPOCH_LENGTH block hashes.
+func (a *ActiveState) RecentBlockHashes() [][32]byte {
+	var blockhashes [][32]byte
+	for _, hash := range a.data.RecentBlockHashes {
+		blockhashes = append(blockhashes, common.BytesToHash(hash))
+	}
+	return blockhashes
+}
+
+// IsVoteCacheEmpty returns false if vote cache of an input block hash doesn't exist.
+func (a *ActiveState) isVoteCacheEmpty(blockHash [32]byte) bool {
+	_, ok := a.blockVoteCache[blockHash]
+	return ok
+}
+
+// GetBlockVoteCache returns the entire set of block vote cache.
+func (a *ActiveState) GetBlockVoteCache() map[[32]byte]*VoteCache {
+	return a.blockVoteCache
+}
+
+func (a *ActiveState) calculateNewAttestations(add []*pb.AttestationRecord, lastStateRecalc uint64) []*pb.AttestationRecord {
+	existing := a.data.PendingAttestations
+	update := []*pb.AttestationRecord{}
+	for i := 0; i < len(existing); i++ {
+		if existing[i].GetSlot() >= lastStateRecalc {
+			update = append(update, existing[i])
+		}
+	}
+	update = append(update, add...)
+
+	return update
+}
+
+func (a *ActiveState) calculateNewBlockHashes(block *Block, parentSlot uint64) ([][]byte, error) {
+	hash, err := block.Hash()
+	if err != nil {
+		return nil, err
+	}
+
+	dist := block.SlotNumber() - parentSlot
+	existing := a.data.RecentBlockHashes
+	update := existing[dist:]
+	for len(update) < 2*params.CycleLength {
+		update = append(update, hash[:])
+	}
+
+	return update, nil
+}
+
+// calculateBlockVoteCache calculates and updates active state's block vote cache.
+func (a *ActiveState) calculateNewVoteCache(block *Block, cState *CrystallizedState) (map[[32]byte]*VoteCache, error) {
+	update := voteCacheDeepCopy(a.GetBlockVoteCache())
+
+	for i := 0; i < len(block.Attestations()); i++ {
+		attestation := block.Attestations()[i]
+
+		parentHashes := a.getSignedParentHashes(block, attestation)
+		attesterIndices, err := cState.getAttesterIndices(attestation)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, h := range parentHashes {
+			// Skip calculating for this hash if the hash is part of oblique parent hashes.
+			var skip bool
+			for _, oblique := range attestation.ObliqueParentHashes {
+				if bytes.Equal(h[:], oblique) {
+					skip = true
```

Comment:
Can you break out of this loop when skip is true? No need to keep checking oblique hashes

Reply:
yeah, that's a good idea